fedbiomed/researcher/federated_workflows/jobs/_training_job.py

```diff
@@ -176,16 +182,23 @@ def execute(self) -> Tuple[Dict, Optional[Dict]]:
 
             self._log_round_info(node=node, training=self._do_training)
 
-            timer[node] = time.perf_counter()
             messages.update({node: TrainRequest(**msg)})  # send request to node
 
-        # Send training request
-        with self._reqs.send(messages, self._nodes) as federated_req:
-            errors = federated_req.errors()
-            replies = federated_req.replies()
-            training_replies = self._get_training_testing_results(replies=replies,
-                                                                  errors=errors,
-                                                                  timer=timer)
+        with self.RequestTimer(self._nodes) as timer:  # compute request time
+            # Send training request
+            with self._reqs.send(messages, self._nodes, self._policies) as federated_req:
+
+                errors = federated_req.errors()
+                replies = federated_req.replies()
+
+        training_replies = self._get_training_results(replies=replies,
+                                                      errors=errors)
+
+        timing_results = self._get_timing_results(replies, timer)
+        # `training_replies` can be empty if there wasnot any replies
+        for node_id in replies:
+            if training_replies.get(node_id):
+                training_replies[node_id].update({'timing': timing_results[node_id]})
```

Comment:
Looking at the original implementation, the function `self._get_training_testing_results` was indented inside the `reqs.send` context. It feels like it should probably be indented similarly, but here it is outside. 
I'm not an expert on the implementation, @srcansiz do you have any thoughts? Is it ok to indent outside or should it be inside the `reqs.send`? 

```suggestion
            training_replies = self._get_training_results(replies=replies,
                                                                                     errors=errors)

            timing_results = self._get_timing_results(replies, timer)
            # `training_replies` can be empty if there wasnot any replies
            for node_id in replies:
                if training_replies.get(node_id):
                    training_replies[node_id].update({'timing': timing_results[node_id]})
```

Reply:
For that I think we were adding extra iterations for nothing. but ok, let's wait for others inputs

Reply:
Hi @sharkovsky, 
Yes you are right. Thank you for pointing this issue. This implementation calculates same `rtime_total` for each node where it should be calculated differently for each node. However, this is not introduced in this PR it was already there because i didn't think about it while I was working on `Requests` redesign. Honestly, I liked @ybouilla 's implementation (context manager for timer) I would say that we can keep it as it is for now and create a new issue to move this timer in `FederatedRequest` and `Request` classes.  Basically, we can have timer for each `Request` that will be accessible through `FedereatedRequest.timings()`.

Reply:
The solution proposed by @ybouilla and @srcansiz is also fine for me.
Thank you all for investigating (and re-explaining it to me this evening - sorry for the fuss).

Just one point to keep in mind for the separate issue: when implementing the moving of the timer in `FederatedRequest` / `Request` pay attention to this detail: the `reply.timing` only exists for `TrainReply` messages. Moreover, it is only checked that it is a `dict` when receiving the message; it is not checked that `reply.timing` has `rtime_training` or `ptime_training` keys. Therefore, as in current implementation we should not try to access these fields without checking their existence (currently we just add a `reply.timing['rtime_training']`

Reply:
I'm a bit confused: 
- for the computation of timing, all is clear and I agree with the current solution + Sergen's proposal
- but the indentation of `training_replies = self._get_training_results(replies=replies, errors=errors)` has also been changed. @srcansiz could you confirm that this is ok?

Reply:
Let's wait for @srcansiz final word, but my understanding is that he explained (and we agreed) on friday that this was (previously) uselessly called inside `with self._reqs.send()` context: not needed as `errors` and `replies` can be called after removing the context, thus OK with both implementations but maybe cleaner with the new one.

Both are OK for me.

Reply:
OK great! 

I didn't know whether it was needed or not and wanted to be sure to avoid mistakes. 

I aggre with @ybouilla and @mvesin that, since the original implementation was redundant, this one is better.

Reply:
Hi @sharkovsky, 
It seems ok to me unless we want to `rtime_total` covers the time for extracting training replies which may be not necessary.